operator/pkg/controllers/infrastructure/v1alpha1/controller/etcd.go

```diff
@@ -0,0 +1,348 @@
+	cert, key, err := pkiutil.KeyAndCert(certConfig.Config, certConfig.caCert, certConfig.caKey, certConfig.isCA)
+	if err != nil {
+		return nil, nil, fmt.Errorf("creating certificate authority, %w,", err)
+	}
+	certBytes := pem.EncodeToMemory(&pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: cert.Raw,
+	})
+	keyBytes := pem.EncodeToMemory(&pem.Block{
+		Type:  "RSA PRIVATE KEY",
+		Bytes: x509.MarshalPKCS1PrivateKey(key.(*rsa.PrivateKey)),
+	})
+	if err := e.createSecret(ctx, certBytes, keyBytes, controlPlane, certConfig); err != nil {
+		return nil, nil, fmt.Errorf("creating root ca secret, %w", err)
+	}
+	zap.S().Infof("[%s] successfully created cert and key %s", controlPlane.ClusterName(), certConfig.name)
+	return cert, key, nil
+}
+
+// createSecret will store root CA and key as the kubernetes secret
+func (e *etcdProvider) createSecret(ctx context.Context, cert, key []byte, controlPlane *v1alpha1.ControlPlane, certConfig *CertConfig) error {
+	secret := &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      certConfig.name,
+			Namespace: controlPlane.NamespaceName(),
+			OwnerReferences: []metav1.OwnerReference{{
+				APIVersion: controlPlane.APIVersion,
+				Name:       controlPlane.Name,
+				Kind:       controlPlane.Kind,
+				UID:        controlPlane.UID,
+			}},
+		},
+		Type: v1.SecretTypeTLS,
+		Data: map[string][]byte{
+			"tls.crt": cert,
+			"tls.key": key,
+		},
+	}
+	if err := e.kubeClient.Create(ctx, secret); err != nil {
+		return err
+	}
+	return nil
+}
+
+func etcdServerPortNameFor(clusterName string) string {
+	return fmt.Sprintf("etcd-server-ssl-%s", clusterName)
+}
+
+func etcdClientPortNameFor(clusterName string) string {
+	return fmt.Sprintf("etcd-client-ssl-%s", clusterName)
+}
+
+func etcdServiceNameFor(clusterName string) string {
+	return fmt.Sprintf("%s-etcd", clusterName)
+}
+
+func etcdCASecretNameFor(clusterName string) string {
+	return fmt.Sprintf("%s-etcd-ca", clusterName)
+}
+
+func etcdServerSecretNameFor(clusterName string) string {
+	return fmt.Sprintf("%s-etcd-server", clusterName)
+}
+
+func etcdPeerSecretNameFor(clusterName string) string {
+	return fmt.Sprintf("%s-etcd-peer", clusterName)
+}
+
+func etcdLabelFor(clusterName string) map[string]string {
+	return map[string]string{
+		"app": etcdServiceNameFor(clusterName),
+	}
+}
+
+func namespacedName(namespace, obj string) types.NamespacedName {
+	return types.NamespacedName{Namespace: namespace, Name: obj}
+}
+
+func certListFor(controlPlane *v1alpha1.ControlPlane, caCert *x509.Certificate, caKey crypto.Signer) []*CertConfig {
+	return []*CertConfig{
+		etcdServerCertConfig(controlPlane, caCert, caKey),
+		etcdPeerCertConfig(controlPlane, caCert, caKey),
+	}
+}
+
+func etcdRootCACertConfig(controlPlane *v1alpha1.ControlPlane) *CertConfig {
+	return &CertConfig{
+		isCA: true,
+		name: etcdCASecretNameFor(controlPlane.ClusterName()),
+		Config: &certutil.Config{
```

Comment:
There's a lot of stuttering with these types. Is there a way we can simplify? CertConfig.Config?

Reply:
Will fix this as part of the work for splitting this file into `controller/etcd`